compiler/src/iree/compiler/Codegen/Common/GPU/GPUTensorAlloc.cpp

```diff
@@ -89,6 +92,53 @@ static bool isSharedMemTranspose(AffineMap indexMap) {
 }
 
 namespace {
+/// Swaps bufferization.alloc_tensor with the copied linalg op result when the
+/// linalg op does not use the output initial value during calculation.
+///
+/// This converts the following IR:
+/// ```
+/// %linalg = linalg ... ins(...) outs(...)
+/// %val = bufferization.alloc_tensor() copy(%linalg)
+/// ```
+/// Into
+/// ```
+/// %alloc = bufferization.alloc_tensor()
+/// %val = linalg ... ins(...) outs(%alloc)
+/// ```
+struct SwapAllocTensorPattern final
+    : public OpRewritePattern<bufferization::AllocTensorOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(bufferization::AllocTensorOp allocOp,
+                                PatternRewriter &rewriter) const override {
+    if (!allocOp.getCopy()) return failure();
+    auto linalgOp = allocOp.getCopy().getDefiningOp<linalg::LinalgOp>();
```

Comment:
Maybe do this for all ops that implement destination passing style.

Reply:
Yeah we can use that once a case arises and so we can have a test for it too. Right now this allows us to use `payloadUsesValueFromOperand` from Linalg op interface to optimize it a bit.